### PR TITLE
pkg/trace: remove gogo from stats protobuf

### PR DESCRIPTION
**Pull request by knusbaum**
### What does this PR do?

This commit replaces gogo protobuf with the first-party protobuf generator for Go, and regenerates the protocol buffer code for stats.

### Motivation

Gogo is no longer maintained, so we should move off of it. Also, we want to migrate our proto files into the agent's main proto package along with all the other protobuf stuff, which will require us to use the same code generator as the rest of the agent
(See also #14845 )

### Additional Notes

The command to generate the code are:
```
protoc -I. --go_out=paths=source_relative:. stats.proto
```

These should maybe live in go:generate, however go:generate currently contains msgpack generation commands which clobber our hand-written modifications to the msgpack encoders.

### Possible Drawbacks / Trade-offs

Unfortunately getting rid of gogo required a change in the types (going from []T to []*T for several nested messages), which in turn required some major refactoring across several packages.

There are subtle semantic differences between []T and []*T in some places we have to be careful for, and should review this PR very closely.

### Describe how to test/QA your changes

TODO: Get input from metrics teams on good QA strategy

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/trace/stats/client_stats_aggregator_test.go [hunk 1]
```diff
@@ -227,30 +227,36 @@ func TestTimeShifts(t *testing.T) {
...
-		expectedAggCounts := wrapPayload(agg2Counts(insertionTime, merge1))
-		// map gives random orders post aggregation
-		assert.ElementsMatch(aggCounts.Stats[0].Stats[0].Stats, expectedAggCounts.Stats[0].Stats[0].Stats)
-		aggCounts.Stats[0].Stats[0].Stats = nil
-		expectedAggCounts.Stats[0].Stats[0].Stats = nil
-		assert.Equal(expectedAggCounts, aggCounts)
-		assert.Len(a.buckets, 0)
-	}
-}
+// func TestFuzzCountFields(t *testing.T) {
```
Comment (knusbaum) [inline pkg/trace/stats/client_stats_aggregator_test.go:230]: TODO

Still need to fix this.

File: pkg/trace/agent/agent.go [hunk 2]
```diff
@@ -422,25 +422,25 @@ func (a *Agent) processStats(in pb.ClientStatsPayload, lang, tracerVersion strin
...
+		mergeDuplicates(*in.Stats[i])
 	}
 	return in
 }
 
 func mergeDuplicates(s pb.ClientStatsBucket) {
 	indexes := make(map[stats.Aggregation]int, len(s.Stats))
 	for i, g := range s.Stats {
-		a := stats.NewAggregationFromGroup(g)
+		a := stats.NewAggregationFromGroup(*g)
```
Comment (katiehockman) [inline pkg/trace/agent/agent.go:443]: Would it be safe to just change `stats.NewAggregationFromGroup` to take a pointer instead?

File: pkg/trace/stats/client_stats_aggregator.go [hunk 3]
```diff
@@ -205,6 +205,9 @@ func (b *bucket) aggregateCounts(p pb.ClientStatsPayload) {
 	}
 	for _, s := range p.Stats {
 		for _, sb := range s.Stats {
+			if sb == nil {
+				sb = &pb.ClientGroupedStats{}
+			}
```
Comment (katiehockman) [inline pkg/trace/stats/client_stats_aggregator.go:210]: Isn't this check already happening? https://github.com/DataDog/datadog-agent/blob/ec8f26cc45b1376f8a3e04e2916f194b10bf215d/pkg/trace/stats/client_stats_aggregator.go#L270

File: pkg/trace/writer/stats.go [hunk 4]
```diff
@@ -233,10 +234,10 @@ func (w *StatsWriter) resolveContainerTags(p *pb.ClientStatsPayload) {
 	}
 }
 
-func splitPayloads(payloads []pb.ClientStatsPayload, maxEntriesPerPayload int) []clientStatsPayload {
+func splitPayloads(payloads []*pb.ClientStatsPayload, maxEntriesPerPayload int) []clientStatsPayload {
 	split := make([]clientStatsPayload, 0, len(payloads))
 	for _, p := range payloads {
-		split = append(split, splitPayload(p, maxEntriesPerPayload)...)
+		split = append(split, splitPayload(*p, maxEntriesPerPayload)...)
```
Comment (katiehockman) [inline pkg/trace/writer/stats.go:240]: We should make sure `p` isn't nil before we try to reference it with `*p`

File: pkg/trace/writer/stats.go [hunk 5]
```diff
@@ -196,14 +196,15 @@ func (w *StatsWriter) buildPayloads(sp pb.StatsPayload, maxEntriesPerPayload int
 		nbEntries = 0
 		nbBuckets = 0
 	}
-	for _, p := range split {
+	for _, pa := range split {
+		p := pa
```
Comment (katiehockman) [inline pkg/trace/writer/stats.go:200]: should we check if p is nil here before continuing?

**Review by katiehockman — COMMENTED** [review]
I combed it over to make sure there wouldn't be any potential panics from nil pointer references/dereferences 👍

**Pull request closed without merging**